src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpTraceHelper.cs

```diff
@@ -0,0 +1,252 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Security;
+
+namespace System.Net.Http
+{
+    internal static class WinHttpTraceHelper
+    {
+        private const string WinHtpTraceEnvironmentVariable = "WINHTTPHANDLER_TRACE";
+        private static bool s_TraceEnabled;
+
+        static WinHttpTraceHelper()
+        {
+            string env;
+            try
+            {
+                env = Environment.GetEnvironmentVariable(WinHtpTraceEnvironmentVariable);
+            }
+            catch (SecurityException)
+            {
+                env = null;
+            }
+            
+            if (!string.IsNullOrEmpty(env))
+            {
+                s_TraceEnabled = true;
+            }
+            else
+            {
+                s_TraceEnabled = false;
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool IsTraceEnabled()
+        {
+            if (s_TraceEnabled)
+            {
+                return true;
+            }
+            
+            return false;
+        }
+
+        public static void TraceStatusCallback(string message, IntPtr handle, bool invokeCallback, uint status)
+        {
+            if (!IsTraceEnabled())
+            {
+                return;
+            }
+
+            Debug.WriteLine(
+                "{0}: handle=0x{1:X}, {2}, {3}",
+                message,
+                handle,
+                invokeCallback ? "processing" : "skipping",
+                GetStringFromInternetStatus(status));
+        }
+
+        public static void TraceAsyncError(string message, Interop.WinHttp.WINHTTP_ASYNC_RESULT asyncResult)
+        {
+            if (!IsTraceEnabled())
+            {
+                return;
+            }
+
+            uint apiIndex = (uint)asyncResult.dwResult.ToInt32();
+            uint error = asyncResult.dwError;
+
+            Debug.WriteLine(
+                "{0}: api={1}, error={2}({3}) \"{4}\"",
+                message,
+                GetNameFromApiIndex(apiIndex),
+                GetNameFromError(error),
+                error,
+                WinHttpException.GetErrorMessage((int)error));
+        }
+
+        public static void Trace()
+        {
+            if (!IsTraceEnabled())
+            {
+                return;
+            }
+        }
+
+        private static string GetNameFromApiIndex(uint index)
+        {
+            switch (index)
+            {
+                case Interop.WinHttp.API_RECEIVE_RESPONSE:
```

Comment:
https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/interop-guidelines.md

>Using enums instead of partial, static classes can lead to needing lots of casts at call sites and can cause problems if such a type needs to be split across multiple files (enums can't currently be partial). However, enums can be valuable in making it clear in a DllImport signature what values are permissible. Enums may be used in limited circumstances where these aren't concerns: the full set of values can be represented in the enum, and the interop signature can be defined to use the enum type rather than the underlying integral type.